### PR TITLE
.Net: Add ThoughtSignature support for Gemini function calling

### LINKED ISSUE #13417: .Net: Bug: Gemini function calling fails with HTTP 400 when thinking is enabled (missing thoughtSignature support)
**Describe the bug**

When using Gemini 2.5 Pro or Gemini 3 Pro with thinking enabled (`thinkingConfig`), function calling fails with HTTP 400:

```
Function call is missing a thought_signature in functionCall parts
```

The Gemini connector does not preserve `thoughtSignature` tokens from API responses, which are mandatory for function calling when thinking is enabled.

**To Reproduce**

1. Configure Gemini connector with Gemini 3 Pro (or 2.5 Pro with `thinkingConfig` enabled)
2. Register a kernel function/tool
3. Send a prompt that triggers function calling
4. Model returns function call with `thoughtSignature` in response
5. Connector builds follow-up request with function result
6. See HTTP 400 error: "Function call is missing a thought_signature in functionCall parts"

**Expected behavior**

Function calling should work correctly. The connector should:
1. Capture `thoughtSignature` tokens from Gemini responses
2. Include them in subsequent requests when sending function results

This is required by the [Gemini API Thought Signatures documentation](https://ai.google.dev/gemini-api/docs/thought-signatures).

**Screenshots**

N/A - API error response.

**Platform**

- Language: C#
- Source: NuGet package `Microsoft.SemanticKernel.Connectors.Google` (latest) / main branch
- AI model: Google Gemini 2.5 Pro, Gemini 3 Pro (any model with thinking enabled)
- IDE: Visual Studio / VS Code
- OS: Windows, Linux, Mac (all platforms affected)

**Additional context**

Per [Google's documentation](https://ai.google.dev/gemini-api/docs/thought-signatures):

> "When a model generates a response that includes thinking, it provides an encrypted representation of its thought process called a thought signature. This signature must be sent back with subsequent turns to maintain the reasoning context."

**Current state:**
-  `thinkingConfig` support added in PR #13383
-  `thoughtSignature` capture from responses: NOT implemented
-  `thoughtSignature` restore in requests: NOT implemented

**API Behavior:**
- For function calls: `thoughtSignature` is **mandatory** (HTTP 400 without it)
- For text responses: `thoughtSignature` is **recommended** for optimal performance
- Parallel function calls: Only the first `functionCall` part has a `thoughtSignature`

**Related:** #13360 (IncludeThoughts / Thought Signatures)

**Workaround:** Disable thinking (loses reasoning quality benefits) or use custom Gemini client.

### PR DESCRIPTION
**Pull request by Cozmopolit**
### Motivation and Context

Fixes #13417

Gemini 2.5/3 Pro with thinking enabled requires `thoughtSignature` tokens for function calling. Without this fix, function calling fails with HTTP 400:

```
Function call is missing a thought_signature in functionCall parts
```

This PR implements full ThoughtSignature support as specified in [Google's Thought Signatures documentation](https://ai.google.dev/gemini-api/docs/thought-signatures).

Full Specification used for this Fix is attached.

### Description

This PR adds `thoughtSignature` support to the Gemini connector:

**Model Changes:**
- Add `ThoughtSignature` property to `GeminiPart` for JSON serialization
- Add `ThoughtSignature` property to `GeminiFunctionToolCall` to carry signature through SK abstractions
- Add `ThoughtSignature` property to `GeminiMetadata` for text response signatures
- Add constructor overload to `GeminiChatMessageContent` accepting `GeminiPart[]` for signature capture

**Response Processing:**
- Capture `thoughtSignature` from function call responses into `GeminiFunctionToolCall`
- Capture `thoughtSignature` from text responses into `GeminiMetadata`

**Request Building:**
- Restore `thoughtSignature` to function call parts when building requests from chat history
- Restore `thoughtSignature` to text parts for model messages

**Bug Fixes:**
- Fix empty ToolCalls list incorrectly treated as containing tools (streaming path)
- Fix grammar: "an Gemini" → "a Gemini" in XML doc comment

**Test Coverage:**
- 29 new unit tests covering serialization, capture, restore, and edge cases
- All 376 tests pass

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone lately  😄

[semantic-kernel-gemini-connector-pr.md](https://github.com/user-attachments/files/24143060/semantic-kernel-gemini-connector-pr.md)

### CODE REVIEW

**Comment by Cozmopolit** [comment]
@microsoft-github-policy-service agree

**Comment by TommyAgami92** [comment]
Any forecast when this will be merged? we cannot use Gemini 3 flash with functions.

**Comment by roldengarm** [comment]
@Cozmopolit have you tested this with Gemini 3 Flash? 
I ran into the bug that functions were not called with Gemini 3; only with Gemini 2. I reproduced it with a console app, see [here](https://github.com/microsoft/semantic-kernel/issues/13400#issuecomment-3716034170).
When I merge in this branch, the problem still exists, it doesn't call functions. I've let Claude fix it quickly, and it was related to thoughts.

See the fix [here](https://github.com/roldengarm/semantic-kernel/commit/01c1531eb18bc47d7b86300aae13539a7dc939fb#diff-eaa7a6f11158c6027cb77ab5c8ebae4f4e69b2ff23ccf94d3f64dc51f56d0421).

**Comment by TommyAgami92** [comment]
@roldengarm did it fix the issue? you were able to call functions with gemini 3 flash using this fix?

**Comment by roldengarm** [comment]
@TommyAgami92 yes, with the fix I did, it works fine with function calling and thoughts enabled. 

If you want to check out for yourself, see [this branch ](https://github.com/roldengarm/semantic-kernel/tree/feature/13400-13418-sample-fix) and instructions to run [here](https://github.com/microsoft/semantic-kernel/issues/13400#issuecomment-3716034170)

It has all changes from @Cozmopolit  from this branch as well, so I can't raise a PR (yet). But, at this stage I think this PR is not complete as it doesn't work for me with Gemini3 without my fixes.

**Comment by TommyAgami92** [comment]
@roldengarm thanks a lot , will check it out!

**Comment by Cozmopolit** [comment]
@roldengarm 

Will check out your branch and amend this PR tommorow. Thanks!

**Comment by roldengarm** [comment]
@Cozmopolit just checking if you had a chance to look into this? Please bear in mind my code change was quickly tested and I haven't added/updated automated tests, but I'm happy to add that if you confirm the change is OK.

**Comment by Cozmopolit** [comment]
@roldengarm   There is a PR to _your_ branch&fork waiting for your comment.

**Comment by Cozmopolit** [comment]
@roldengarm

While I try to get this right and we wait for MS to react here, you might want to consider:
https://github.com/Cozmopolit/SemanticKernel.Connectors.GeminiDotnet

**Comment by roldengarm** [comment]
> @roldengarm There is a PR to _your_ branch&fork waiting for your comment.

Thanks @Cozmopolit , looks good, approved 👍

**Comment by roldengarm** [comment]
@rogerbarreto as you reviewed my previous [PR regarding Gemini & Thinking](https://github.com/microsoft/semantic-kernel/pull/13383), would this also be a PR you could review please?
right now, function calling with Gemini 3+ doesn't work

**Comment by markwallace-microsoft** [comment]
@Cozmopolit there is a failing unit test

```
[xUnit.net 00:00:00.78]     SemanticKernel.Connectors.Google.UnitTests.Core.Gemini.Clients.GeminiChatClientFunctionCallingTests.ChatClientShouldReceiveFunctionCallsInResponseAsync [FAIL]
[xUnit.net 00:00:00.78]       Assert.NotEmpty() Failure: Collection was empty
[xUnit.net 00:00:00.78]       Stack Trace:
[xUnit.net 00:00:00.78]         /home/runner/work/semantic-kernel/semantic-kernel/dotnet/src/Connectors/Connectors.Google.UnitTests/Core/Gemini/Clients/GeminiChatClientFunctionCallingTests.cs(114,0): at SemanticKernel.Connectors.Google.UnitTests.Core.Gemini.Clients.GeminiChatClientFunctionCallingTests.ChatClientShouldReceiveFunctionCallsInResponseAsync()
```

**Comment by markwallace-microsoft** [comment]
It looks like the tool calls are not being correctly converted to `Microsoft.SemanticKernel.FunctionCallContent` and end up missing from the chat response.

One of the `GeminiChatMessageContent` constructors doesn't populate Items collection with FunctionCallContent for compatibility with FunctionChoiceBehavior.

File: dotnet/src/Connectors/Connectors.Google/Models/Gemini/GeminiChatMessageContent.cs [hunk 1]
```diff
@@ -191,6 +191,29 @@ internal GeminiChatMessageContent(
             .Where(p => p.FunctionCall is not null)
             .Select(part => new GeminiFunctionToolCall(part))
             .ToList();
+
+        // Also populate Items collection with FunctionCallContent for compatibility with FunctionChoiceBehavior
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.Google/Models/Gemini/GeminiChatMessageContent.cs]: This looks like duplicated code, can this be extracted into a private method

**Review by markwallace-microsoft — APPROVED** [review]
